### PR TITLE
node: [WIP] Minimal modes

### PR DESCRIPTION
**Pull request by Eric-Warehime**
## Summary

Introduces "modes" that can be chosen during startup to limit and/or extend the set of functionality that a node (and its algod APIs) has. This currently includes:
* participating (the same as the existing `AlgorandFullNode`)
* non participating (removes participation/agreement and txn broadcasting)
* data (adds APIs from #4658 in order to expose state deltas between blocks)

## Test Plan

Existing tests should all pass. Will be adding more tests in various places including (list subject to change)
**unit tests**
- [ ] handlers
- [ ] node
**integ tests**
- [ ] e2e tests for different node types

### CODE REVIEW

**Comment by Eric-Warehime** [comment]
Catchup service is outputting a bunch of log files along these lines when no-oping block fetches. Not terrible, but it is creating a bunch of noise in the log files.
```
{"Context":"sync","file":"service.go","function":"github.com/algorand/go-algorand/catchup.(*Service).pipelineCallback.func1","level":"info","line":442,"msg":"pipelineCallback(4380291): did not fetch or write the block","name":"","time":"2022-11-17T08:48:30.299944-08:00"}
```

**Comment by Eric-Warehime** [comment]
One issue: invalid round provided to the get deltas API returns a 500 with the following log:
```
{
  "file": "utils.go",
  "function": "github.com/algorand/go-algorand/daemon/algod/api/server/v2.returnError",
  "level": "info",
  "line": 44,
  "msg": "round 4380280 too high: dbRound 4380271, deltas 8",
  "time": "2022-11-17T08:54:18.520478-08:00"
}
```

Valid API requests look as expected. The block [on testnet](https://testnet.algoexplorer.io/block/4380279) has no txns, but we do see the rewards pool account showing up in the deltas.
```
{
  "catchpoint": "",
  "catchpoint-acquired-blocks": 0,
  "catchpoint-processed-accounts": 0,
  "catchpoint-total-accounts": 0,
  "catchpoint-total-blocks": 0,
  "catchpoint-verified-accounts": 0,
  "catchup-time": 0,
  "last-catchpoint": "",
  "last-round": 4380279,
  "last-version": "https://github.com/algorandfoundation/specs/tree/4a9db6a25595c6fd097cf9cc137cc83027787eaa",
  "next-version": "https://github.com/algorandfoundation/specs/tree/4a9db6a25595c6fd097cf9cc137cc83027787eaa",
  "next-version-round": 4380280,
  "next-version-supported": true,
  "stopped-at-unsupported-round": false,
  "time-since-last-round": 665909988084
}
```
```
curl --get http://localhost:8080/v2/ledger/sync
{"round":4380276}
```
```
curl --get http://localhost:8080/v2/deltas/4380279
{
  "accounts": [
    {
      "account-data": {
        "address": "7777777777777777777777777777777777777777777777777774MSJUVU",
        "amount": 8164245494,
        "amount-without-pending-rewards": 7958986206,
        "apps-total-schema": {
          "num-byte-slice": 0,
          "num-uint": 0
        },
        "min-balance": 100000,
        "pending-rewards": 205259288,
        "reward-base": 0,
        "rewards": 0,
        "round": 4380279,
        "status": "Not Participating",
        "total-apps-opted-in": 0,
        "total-assets-opted-in": 0,
        "total-created-apps": 0,
        "total-created-assets": 0
      },
      "address": "7777777777777777777777777777777777777777777777777774MSJUVU"
    }
  ],
  "apps": null,
  "assets": null,
  "kv-deltas": null
}
```

**Review by jannotti — COMMENTED** [review]
I'm not sure I like the approach of trying to bin nodes into different kinds. It seems more straight-forward, both conceptually and in code, to enable or disable certain features. For example, why can't an NPN node broadcast transactions?  That's not "participation" in the sense of voting (nor in the *other* meaning that we elsewhere use for "nonparticipating" - to not receive rewards).  When we name concepts like this, we run into a lot of trouble defining them.

In code, this is also leading to a proliferation of embedded interfaces that seem to be trying to implement inheritance.  I don't think that works well in Go. I see, for example, that it seems to have forced `LedgerForAPI` in the ledger package, but I expect Go interfaces to be declared by consumers of the interface, not by implementors.

**Comment by winder** [comment]
tl;dr - I think there are a lot of good ideas here that would make our code more robust, but maybe this PR is too ambitious.

@jannotti I think the main motivation here is that the node and the services it manages do not have clean feature flags like you described. As more feature are added, it becomes more difficult to consider the different combinations of features.

One perfect example of this is the `SetCatchpointCatchupMode`. Currently the node operates in one mode, so it turns everything on or off. After this change some of the services may not exist, so how do you go about turning them off and on? Even without this change that code is already brittle because the services must be stopped/started in the correct order (implementation details leaked into the node).

The other big problem is that the node complicated. At the surface, the node object seems like a thin wrapper over a bunch of services that do the work. After peeling back the layers a bit it's clearly much more to it than that. By breaking apart these services it forces us to consider them individually rather than blobs of code that overlap each other in subtle ways.

Take for example is `network.NewWebsocketNetwork`, which takes the node itself as an argument in order to provide the `NodeInfo` interface. Another example is `node.mu` which is used during startup and shutdown only, and carefully worked around during the main loop.

@cce had previously suggested a scheme where we stub out the "unused" services with dummy implementations. That way the full node wouldn't really know or care what they were doing. On the other hand, if we had unit tests that used stubbed out services we could miss regressions that only exist once the real services are interacting with one another.

All that said, this is a very large PR with a lot of changes to critical code paths. I think it would be hard to accept it as-is.

**Comment by Eric-Warehime** [comment]
Thanks for the feedback. I'm going to rework this based on all the conversations we've had recently.
